starlark/library.go

```diff
@@ -2218,34 +2222,89 @@ func set_clear(_ *Thread, b *Builtin, args Tuple, kwargs []Tuple) (Value, error)
 
 // https://github.com/google/starlark-go/blob/master/doc/spec.md#set·difference.
 func set_difference(_ *Thread, b *Builtin, args Tuple, kwargs []Tuple) (Value, error) {
-	// TODO: support multiple others: s.difference(*others)
-	var other Iterable
-	if err := UnpackPositionalArgs(b.Name(), args, kwargs, 0, &other); err != nil {
-		return nil, err
+	accum := b.Receiver().(*Set).clone()
+	if err := forIterableArgsDo(accum.DifferenceUpdate, args, kwargs); err != nil {
+		return nil, nameErr(b, err)
 	}
-	iter := other.Iterate()
-	defer iter.Done()
-	diff, err := b.Receiver().(*Set).Difference(iter)
-	if err != nil {
+	return accum, nil
+}
+
+// https://github.com/google/starlark-go/blob/master/doc/spec.md#set·difference_update.
+func set_difference_update(_ *Thread, b *Builtin, args Tuple, kwargs []Tuple) (Value, error) {
+	recv := b.Receiver().(*Set)
+	if err := recv.ht.checkMutable("delete from"); err != nil {
+		// even if args is empty
 		return nil, nameErr(b, err)
 	}
-	return diff, nil
+	if err := forIterableArgsDo(recv.DifferenceUpdate, args, kwargs); err != nil {
+		return nil, nameErr(b, err)
+	}
+	return None, nil
 }
 
 // https://github.com/google/starlark-go/blob/master/doc/spec.md#set_intersection.
 func set_intersection(_ *Thread, b *Builtin, args Tuple, kwargs []Tuple) (Value, error) {
-	// TODO: support multiple others: s.difference(*others)
+	accum := b.Receiver().(*Set).clone()
+	// TODO: use an intersection primitive that preserves lhs iteration order
```

Comment:
We should address this in this CL. I think we should explicitly guarantee the insertion order of all these operations.

Reply:
Done.